### PR TITLE
Allow retries for registry git clone & `download_verify` to make CI more robust

### LINKED ISSUE #3238: MacOS git registry CI failing: `GitError(Code:ERROR, Class:Net, SecureTransport error: connection closed via error)`
```
Depot setup: Error During Test at /Users/runner/work/Pkg.jl/Pkg.jl/test/new.jl:28
  Got exception outside of a @test
  failed to clone from https://github.com/JuliaRegistries/General.git, error: GitError(Code:ERROR, Class:Net, SecureTransport error: connection closed via error)
  Stacktrace:
    [1] pkgerror(msg::String)
      @ Pkg.Types ~/work/Pkg.jl/Pkg.jl/src/Types.jl:68
    [2] clone(io::Base.BufferStream, url::String, source_path::String; header::Nothing, credentials::LibGit2.CachedCredentials, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ Pkg.GitTools ~/work/Pkg.jl/Pkg.jl/src/GitTools.jl:126
    [3] kwcall(::NamedTuple{(:credentials,), Tuple{LibGit2.CachedCredentials}}, ::typeof(Pkg.GitTools.clone), io::Base.BufferStream, url::String, source_path::String)
      @ Pkg.GitTools ~/work/Pkg.jl/Pkg.jl/src/GitTools.jl:88
    [4] (::Main.PkgTestsOuter.PkgTestsInner.Utils.var"#1#3")(creds::LibGit2.CachedCredentials)
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:36
    [5] shred!(f::Main.PkgTestsOuter.PkgTestsInner.Utils.var"#1#3", x::LibGit2.CachedCredentials)
      @ Base ./secretbuffer.jl:193
    [6] init_reg()
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:35
    [7] isolate(fn::Main.PkgTestsOuter.PkgTestsInner.NewTests.var"#1#12"; loaded_depot::Bool, linked_reg::Bool)
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:58
    [8] isolate(fn::Function)
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:46
    [9] macro expansion
      @ ~/work/Pkg.jl/Pkg.jl/test/new.jl:29 [inlined]
   [10] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1496 [inlined]
   [11] top-level scope
      @ ~/work/Pkg.jl/Pkg.jl/test/new.jl:29
   [12] include(mod::Module, _path::String)
      @ Base ./Base.jl:450
   [13] include
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:9 [inlined]
   [14] macro expansion
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:70 [inlined]
   [15] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1584 [inlined]
   [16] macro expansion
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:52 [inlined]
   [17] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:[149](https://github.com/JuliaLang/Pkg.jl/actions/runs/3316170742/jobs/5477648317#step:7:150)6 [inlined]
   [18] (::Main.PkgTestsOuter.PkgTestsInner.var"#1#2")()
      @ Main.PkgTestsOuter.PkgTestsInner ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:51
   [19] with_logstate(f::Function, logstate::Any)
      @ Base.CoreLogging ./logging.jl:514
   [20] with_logger(f::Function, logger::Base.CoreLogging.NullLogger)
      @ Base.CoreLogging ./logging.jl:626
   [21] top-level scope
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:49
   [22] include(fname::String)
      @ Base.MainInclude ./client.jl:478
   [23] top-level scope
      @ none:6
   [24] eval
      @ ./boot.jl:370 [inlined]
   [25] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:280
   [26] _start()
      @ Base ./client.jl:522
  
  caused by: GitError(Code:ERROR, Class:Net, SecureTransport error: connection closed via error)
  Stacktrace:
    [1] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/LibGit2/src/error.jl:111 [inlined]
    [2] clone(repo_url::SubString{String}, repo_path::String, clone_opts::LibGit2.CloneOptions)
      @ LibGit2 ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/LibGit2/src/repository.jl:459
    [3] clone(repo_url::SubString{String}, repo_path::String; branch::String, isbare::Bool, remote_cb::Ptr{Nothing}, credentials::LibGit2.CachedCredentials, callbacks::Dict{Symbol, Tuple{Ptr{Nothing}, Any}})
      @ LibGit2 ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/LibGit2/src/LibGit2.jl:583
    [4] clone
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/LibGit2/src/LibGit2.jl:556 [inlined]
    [5] clone(io::Base.BufferStream, url::String, source_path::String; header::Nothing, credentials::LibGit2.CachedCredentials, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ Pkg.GitTools ~/work/Pkg.jl/Pkg.jl/src/GitTools.jl:115
    [6] kwcall(::NamedTuple{(:credentials,), Tuple{LibGit2.CachedCredentials}}, ::typeof(Pkg.GitTools.clone), io::Base.BufferStream, url::String, source_path::String)
      @ Pkg.GitTools ~/work/Pkg.jl/Pkg.jl/src/GitTools.jl:88
    [7] (::Main.PkgTestsOuter.PkgTestsInner.Utils.var"#1#3")(creds::LibGit2.CachedCredentials)
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:36
    [8] shred!(f::Main.PkgTestsOuter.PkgTestsInner.Utils.var"#1#3", x::LibGit2.CachedCredentials)
      @ Base ./secretbuffer.jl:193
    [9] init_reg()
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:35
   [10] isolate(fn::Main.PkgTestsOuter.PkgTestsInner.NewTests.var"#1#12"; loaded_depot::Bool, linked_reg::Bool)
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:58
   [11] isolate(fn::Function)
      @ Main.PkgTestsOuter.PkgTestsInner.Utils ~/work/Pkg.jl/Pkg.jl/test/utils.jl:46
   [12] macro expansion
      @ ~/work/Pkg.jl/Pkg.jl/test/new.jl:29 [inlined]
   [13] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1496 [inlined]
   [14] top-level scope
      @ ~/work/Pkg.jl/Pkg.jl/test/new.jl:29
   [15] include(mod::Module, _path::String)
      @ Base ./Base.jl:450
   [16] include
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:9 [inlined]
   [17] macro expansion
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:70 [inlined]
   [18] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1584 [inlined]
   [19] macro expansion
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:52 [inlined]
   [20] macro expansion
      @ ~/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1496 [inlined]
   [21] (::Main.PkgTestsOuter.PkgTestsInner.var"#1#2")()
      @ Main.PkgTestsOuter.PkgTestsInner ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:51
   [22] with_logstate(f::Function, logstate::Any)
      @ Base.CoreLogging ./logging.jl:514
   [23] with_logger(f::Function, logger::Base.CoreLogging.NullLogger)
      @ Base.CoreLogging ./logging.jl:626
   [24] top-level scope
      @ ~/work/Pkg.jl/Pkg.jl/test/runtests.jl:49
   [25] include(fname::String)
      @ Base.MainInclude ./client.jl:478
   [26] top-level scope
      @ none:6
   [27] eval
      @ ./boot.jl:370 [inlined]
   [28] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:280
   [29] _start()
      @ Base ./client.jl:522
```

### PR DESCRIPTION
**Pull request by IanButterworth**
Trying to fix #3238 

cc. @StefanKarpinski

### CODE REVIEW

**Comment by IanButterworth** [comment]
Could this be upstream LibGit2 issues on MacOS?

We're two releases behind. 1.4.3, while 1.4.4 and 1.5 are out https://github.com/libgit2/libgit2/releases

File: src/Registry/Registry.jl [hunk 1]
```diff
@@ -425,12 +428,24 @@ function update(regs::Vector{RegistrySpec} = RegistrySpec[]; io::IO=stderr_f(),
...
+                                    sleep(1)
+                                    @goto merge
+                                elseif e isa LibGit2.GitError && e.code == LibGit2.Error.ENOTFOUND
+                                    push!(errors, (reg.path, "branch origin/$branch not found"))
+                                    @goto done_git
+                                else
+                                    rethrow()
+                                end
+
                             end
```
Comment (IanButterworth) [inline src/Registry/Registry.jl:449]: @StefanKarpinski does this seem reasonable to you?

Comment (StefanKarpinski) [reply src/Registry/Registry.jl:454]: Casual glance but yes, looks good to me 👍🏻

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -61,7 +61,7 @@ jobs:
       - run: julia --project --color=yes --check-bounds=yes -e 'import Pkg; Pkg.build(); Pkg.test(; coverage=true)'
         env:
           JULIA_PKG_SERVER: ${{ matrix.pkg-server }}
-          JULIA_PKG_TEST_QUIET: "true" # "true" is the default. i.e. tests are quiet when this env var isn't set
+          JULIA_PKG_TEST_QUIET: "false" # "true" is the default. i.e. tests are quiet when this env var isn't set
```
Comment (IanButterworth) [inline .github/workflows/test.yml:64]: @DilumAluthge It'd be nice to see that this works at least once before merge. So let's leave this on for now and re-run CI until we see the `@warn "Registry update attempt failed...` log, then turn it off before merge

```suggestion
          JULIA_PKG_TEST_QUIET: "true" # "true" is the default. i.e. tests are quiet when this env var isn't set
```

File: src/Registry/Registry.jl [hunk 3]
```diff
@@ -228,7 +231,8 @@ function download_registries(io::IO, regs::Vector{RegistrySpec}, depot::String=d
                     printpkgstyle(io, :Copied, "registry `$(Base.contractuser(registry.name))` to `$(Base.contractuser(regpath))`")
                     return
                 elseif reg.url !== nothing # clone from url
-                    repo = GitTools.clone(io, reg.url, tmp; header = "registry from $(repr(reg.url))")
+                    # retry to help spurious connection issues, particularly on CI
+                    repo = retry(GitTools.clone, delays = fill(1.0, 10), check=(s,e)->isa(e, LibGit2.GitError))(io, reg.url, tmp; header = "registry from $(repr(reg.url))")
```
Comment (IanButterworth) [inline src/Registry/Registry.jl:235]:
```suggestion
                    repo = retry(GitTools.clone, delays = fill(1.0, 5), check=(s,e)->isa(e, LibGit2.GitError))(io, reg.url, tmp; header = "registry from $(repr(reg.url))")
```

**Comment by IanButterworth** [comment]
Seems to be more stable now. 

It would be nice if `Base.retry` had a mechanism for logging when retries are done.

**Pull request merged**